File: core/src/rustyice/game/Game.java
package rustyice.game;

import box2dLight.RayHandler;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.*;
import rustyice.physics.Collidable;
import rustyice.physics.Collision;
import rustyice.game.tiles.TileMap;
import rustyice.graphics.Camera;
import rustyice.resources.Resources;

import java.util.ArrayList;

public class Game implements ContactListener {
    private static final float UPDATE_RATE = 1/60f;

    private RayHandler rayHandler;
    private Resources resources;
    private World world;
    private Section currentSection;
    private Section sectionToLoad;
    private ArrayList<Collision> collisions;
    private float leftOverTime;

    public Game() {
        world = new World(Vector2.Zero, true);
        world.setContactListener(this);

        rayHandler = new RayHandler(world);
        leftOverTime = 0;

        sectionToLoad = new Section();

        // actors = new ArrayList<>();
        collisions = new ArrayList<>();
    }

    public Section getCurrentSection() {
        return currentSection;
    }

    public RayHandler getRayHandler() {
        return rayHandler;
    }

    private void loadSection(Section section) {
        if (currentSection != null) {
            currentSection.store();
        }

        section.setGame(this);
        section.init();

        currentSection = section;
    }

    public void finishLoadingSection() {
        if (sectionToLoad != null) {
            loadSection(sectionToLoad);
            sectionToLoad = null;
        }
    }

    public void setSectionToLoad(Section sectionToLoad) {
        this.sectionToLoad = sectionToLoad;
    }

    public TileMap getTiles() {
        return currentSection.getTiles();
    }

    public void setResources(Resources resources) {
        this.resources = resources;
    }

    public Resources getResources() {
        return resources;
    }

    public void update(float delta) {
        finishLoadingSection();
        
        leftOverTime += delta;

        while (leftOverTime > 0) {
            leftOverTime -= UPDATE_RATE; 
            world.step(UPDATE_RATE, 6, 2);

            for (Collision col : collisions) {
                if (col.isBegin()) {
                    Collidable collidable = resolveCollidable(col.getThisFixture());
                    if (collidable != null) {
                        collidable.beginCollision(col);
                    }
                } else {
                    Collidable collidable = resolveCollidable(col.getThisFixture());
                    if (collidable != null) {
                        collidable.endCollision(col);
                    }
                }
            }
            collisions.clear();

            currentSection.update(UPDATE_RATE);
        }
    }

    public void render(SpriteBatch batch, Camera camera, int flags) {
        currentSection.render(batch, camera, flags);
    }

    public World getWorld() {
        return world;
    }

    public void dispose() {
        world.dispose();
    }

    public Collidable resolveCollidable(Fixture fixture) {
        return (Collidable) fixture.getUserData();
    }

    @Override
    public void beginContact(Contact contact) {
        collisions.add(new Collision(contact.getFixtureA(), contact.getFixtureB(), true));
        collisions.add(new Collision(contact.getFixtureB(), contact.getFixtureA(), true));
    }

    @Override
    public void endContact(Contact contact) {
        collisions.add(new Collision(contact.getFixtureA(), contact.getFixtureB(), false));
        collisions.add(new Collision(contact.getFixtureB(), contact.getFixtureA(), false));
    }

    @Override
    public void preSolve(Contact contact, Manifold manifold) {

    }

    @Override
    public void postSolve(Contact contact, ContactImpulse contactImpulse) {

    }
}

File: core/src/rustyice/game/Game.kt
package rustyice.game

import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.*
import rustyice.graphics.Camera
import rustyice.graphics.RenderLayer
import rustyice.input.PlayerInput
import rustyice.physics.Collidable
import rustyice.physics.Collision
import java.util.*

class Game: ContactListener {
    private val UPDATE_RATE = 1/60f

    val world: World

    var currentSection: Section? = null
        private set

    private var sectionToLoad: Section? = null
    private val collisions: ArrayList<Collision>
    private var timeAccumulator: Float = 0f

    init {
        world = World(Vector2.Zero, true)
        world.setContactListener(this)

        sectionToLoad = Section()

        // actors = new ArrayList<>();
        collisions = ArrayList()
    }

    fun startLoadingSection(section: Section){
        sectionToLoad = section
    }

    fun loadSectionNow(section: Section){
        startLoadingSection(section)
        finishLoadingSection()
    }

    fun finishLoadingSection() {
        val section = sectionToLoad

        if(section != null) {
            currentSection?.store()

            section.game = this
            section.init()

            currentSection = section
            sectionToLoad = null
        }
    }

    fun update(delta: Float) {
        finishLoadingSection()
        
        timeAccumulator += delta

        while (timeAccumulator > 0) {
            timeAccumulator -= UPDATE_RATE

            world.step(UPDATE_RATE, 6, 2)

            for (col in collisions) {
                if (col.isBegin) {
                    resolveCollidable(col.thisFixture)?.beginCollision(col)
                } else {
                    resolveCollidable(col.thisFixture)?.endCollision(col)
                }
            }
            collisions.clear()

            currentSection?.update(UPDATE_RATE)
        }
    }

    fun render(batch: Batch, camera: Camera, layer: RenderLayer) {
        currentSection?.render(batch, camera, layer)
    }

    fun dispose() {
        currentSection?.store()

        world.dispose()
    }

    fun resolveCollidable(fixture: Fixture): Collidable? {
        return fixture.userData as Collidable?
    }

    override fun beginContact(contact: Contact) {
        collisions.add(Collision(contact.fixtureA, contact.fixtureB, true))
        collisions.add(Collision(contact.fixtureB, contact.fixtureA, true))
    }

    override fun endContact(contact: Contact) {
        collisions.add(Collision(contact.fixtureA, contact.fixtureB, false))
        collisions.add(Collision(contact.fixtureB, contact.fixtureA, false))
    }

    override fun preSolve(contact: Contact, manifold: Manifold) {}
    override fun postSolve(contact: Contact, contactImpulse: ContactImpulse) {}
}